server/services/content-types.js

```diff
@@ -146,38 +214,102 @@ const manageRelations = async (newData, uid, oldVersionId, model) => {
     return newData;
   }
 
-  const updatableRelations = getUpdatableRelations(model);
+  const updatableRelations = reduceArray(getUpdatableRelations(model));
+
   const previousVersion = await strapi.db.query(uid).findOne({
     where: {
       id: oldVersionId,
     },
-    populate: updatableRelations,
+    populate: generatePopulateStatement(updatableRelations),
   });
-  const connects = {};
-  updatableRelations.forEach((rel) => {
-    const prevRel = previousVersion[rel];
-    if (prevRel) {
-      const newDataRel = newData[rel];
-      const newDataConnects = newDataRel.connect;
-      const newDataDisconnects = newDataRel.disconnect;
-      //Connect relations from previous version but only if they were not changed by user in the current version.
-      const idsHandledByEditor = [
-        ...newDataConnects.map((conn) => conn.id),
-        ...newDataDisconnects.map((conn) => conn.id),
-      ];
-      const mergedConnects = [...newDataConnects];
-      const prevRelIds = Array.isArray(prevRel)
-        ? prevRel.map((rel) => rel.id)
-        : [prevRel.id];
-      prevRelIds.forEach((pid) => {
-        if (!idsHandledByEditor.includes(pid)) {
-          mergedConnects.push(pid);
+
+  const mergeConnections = (newDataRel, prevRel) => {
+    const newDataConnects = newDataRel.connect;
+    const newDataDisconnects = newDataRel.disconnect;
+
+    //Connect relations from previous version but only if they were not changed by user in the current version.
+    const idsHandledByEditor = [
+      ...newDataConnects.map((conn) => conn.id),
+      ...newDataDisconnects.map((conn) => conn.id),
+    ];
+
+    const mergedConnects = [...newDataConnects];
+
+    const prevRelIds = Array.isArray(prevRel)
+      ? prevRel.map((rel) => rel.id)
+      : [prevRel.id];
+    prevRelIds.forEach((pid) => {
+      if (!idsHandledByEditor.includes(pid)) {
+        mergedConnects.push({ id: pid });
+      }
+    });
+
+    return mergedConnects;
+  };
+
+  const updateRelations = (updateableRels, previous, parent = null) => {
+    updateableRels.forEach((relation) => {
+      if (typeof relation === "string") {
+        if (previous) {
+          if (!_.isArray(previous)) {
+            const prevRel = previous[relation] ?? undefined;
+
+            if (prevRel) {
+              const newDataRel = parent
```

Comment:
I'm struggling with how to correct this and attach the new data for if the value is more than one level down. e.g. the relation, from the top level would be, `contentType -> dynamicZone -> component -> relation` and therefore setting `newData[parent][relation]` results in an error because it skips a level

Reply:
I thought I was on to something with these functions:
```js
const setDeepValue = (obj, value, path) => {
  let index = 0;
  for (index; index < path.length; index++) {
    obj = obj[path[index]];
  }
  obj[path[index]] = value;
};

const getDeepValue = (obj, path) => {
  for (const index of path) {
    obj = obj[path[index]];
    return obj;
  }
};
```

But haven't managed to get it working for all circumstances with an implementation like
```js
const newDataRel = getDeepValue(newData, [...path, relation]);
```
And don't want to push a half working commit

Reply:
I understand. So did you manage to fix the getUpdateStatement fnc? It would be great for me to have that part so I could debug this one.

Reply:
Sorry I did not get to you quicker. I had to finish work on other project before Christmas + fixed another bug on this plugin. I should be free to deep into this one tomorrow.

Reply:
> I understand. So did you manage to fix the getUpdateStatement fnc? It would be great for me to have that part so I could debug this one.

Sadly not yet, I just hard-coded a data structure that fit with my use case

Reply:
Hey, glad to be back after the holiday. I think I adjusted the populate statement generator. It works for my data structure. I think I will now encouter the problem you have described and will try to fix it.

`const generatePopulateStatement = (relations) => {
  const result = {};

  relations.forEach((item) => {
    if (typeof item === "object" && !Array.isArray(item)) {
      const key = Object.keys(item)[0];
      const innerObj = {};

      item[key].forEach((innerItem) => {
        if (typeof innerItem === "object") {
          const key = Object.keys(innerItem)[0];
          innerObj[key] = {
            populate: {},
          };
          innerItem[key].forEach((element) => {
            innerObj[key]["populate"][element] = true;
          });
        } else {
          innerObj[innerItem] = true;
        }
      });

      result[key] = { populate: innerObj };
    } else if (Array.isArray(item)) {
      const key = Object.keys(item)[0];
      const innerObj = {};

      item[key].forEach((innerItem) => {
        innerObj[innerItem] = true;
      });

      result[key] = { populate: innerObj };
    } else {
      result[item] = true;
    }
  });

  return result;
};`

Reply:
@ShipleyC95 So I tried implementing it and I think I got close. However it still does not work :/ Morover, when there are multiple types of components in the dz it sometimes deletes one of them. 

You can check what I have tried in https://github.com/notum-cz/strapi-plugin-content-versioning/tree/feature/dz-relations-om

Reply:
Thanks for the update @omikulcik. Unfortunately, your generate method doesn't work for me, I end up with one of the keys being "[Object object]" but I will look into it and that branch you've posted there

Reply:
Thank you for testing it @ShipleyC95. Working with this polymorphic data structure seems quite complex to me but hopefully we can somehow solve this. I will get into working on this again during this week.